docs/DesignDocs/shared_memory.md

```diff
@@ -0,0 +1,192 @@
+could pass any pointer to untrusted memory. For example:
+
+```c++
+    static unsigned char buf[4096];
+
+    share_buffer_ecall(buf, sizeof(buf));
+```
+
+But this application is not portable to TrustZone since **buf** is not visible
+within the enclave. To make this work on TrustZone, the buffer should be
+explicitly allocated from a shared memory pool as follows.
+
+```c++
+    const size_t n = 4096;
+    unsigned char* buf;
+
+    if (oe_allocate_shared_memory(enclave, n, &buf) != OE_RESULT_OK)
+    {
+        /* Error */
+    }
+
+    memset(buf, 0, n);
+
+    share_buffer_ecall(buf, n);
+```
+
+This program fragment will work on both SGX and TrustZone. The application
+will eventually free this buffer by passing it to **oe_free_shared_memory()**.
+
+The enclave's implementation of **share_buffer_ecall()** should verify that
+the **untrusted_memory** is fully contained within shared memory as shown
+below.
+
+```c++
+void share_buffer_ecall(void* untrusted_memory, size_t untrusted_memory_size)
+{
+    if (!oe_is_shared_memory(untrusted_memory, untrusted_memory_size)
+    {
+        /* Error! */
+    }
+    ...
+}
+```
+
+Specification
+-------------
+
+This design introduces the following functions on the host side:
+
+```c++
+    oe_result_t oe_allocate_shared_memory(
+        oe_enclave_t* enclave,
+        size_t size,
+        void** ptr);
+
+    oe_result_t oe_free_shared_memory(
+        oe_enclave_t* enclave,
+        void* ptr);
+
+    bool oe_is_shared_memory(
+        oe_enclave_t* enclave,
+        const void* ptr,
+        size_t size);
+
+    typedef enum _oe_memory_type
+    {
+        /* !oe_is_within_enclave() && !oe_is_shared_memory() */
+        OE_MEMORY_TYPE_UNKNOWN,
+
+        /* oe_is_within_enclave() */
+        OE_MEMORY_TYPE_TRUSTED,
+
+        /* oe_is_shared_memory() */
+        OE_MEMORY_TYPE_SHARED,
+    }
+    oe_memory_type_t;
+
+    oe_result_t oe_get_memory_type(
+        oe_enclave_t* enclave,
+        const void* ptr,
+        size_t size,
+        oe_memory_type_t* type);
+```
+
+The same functions are available on the enclave side but without the enclave
+parameter.
+
+```c++
+    oe_result_t oe_allocate_shared_memory(
+        size_t size,
+        void** ptr);
+
+    oe_result_t oe_free_shared_memory(
+        void* ptr);
+
+    bool oe_is_shared_memory(
+        const void* ptr,
+        size_t size);
+
+    typedef enum _oe_memory_type
+    {
+        /* !oe_is_within_enclave() && !oe_is_shared_memory() */
+        OE_MEMORY_TYPE_UNKNOWN,
+
+        /* oe_is_within_enclave() */
+        OE_MEMORY_TYPE_TRUSTED,
+
+        /* oe_is_shared_memory() */
+        OE_MEMORY_TYPE_SHARED,
+    }
+    oe_memory_type_t;
+
+    oe_result_t oe_get_memory_type(
+        const void* ptr,
+        size_t size,
+        oe_memory_type_t* type);
+```
+
+Assumptions:
+
+- All shared memory is read-write (non-executable). No mechanism is provided
+  for allocating read-only memory or executable memory.
```

Comment:
This is probably a `nice to have` in the future. An enclave function that writes to the shared memory deserves more security scrutinies than a function that just read it.